### PR TITLE
layers: Add VK_QCOM_tile_shading dispatch VUIDs (2/2)

### PR DESCRIPTION
**Pull request by ZZKQcom**
# Overview
Adds validation for three `VK_QCOM_tile_shading` dispatch-only VUIDs:
- `VUID-vkCmdDispatch-aspectMask-10673`
- `VUID-vkCmdDispatch-None-10675`
- `VUID-vkCmdDispatch-None-10676`

 Introduces `tile_attachments_written_to` (per-subpass set of `ImageView*`) populated by `RecordDrawTileAttachmentsWrittenTo()` and `RecordDispatchTileAttachmentsWrittenTo()` after each `draw/dispatch` command.

# Unit tests

| Test | VUID |
|------|------|
| `DispatchInsidePerTileExecutionModelButUseAspectDepthStencilMask` | 10673 |
| `DispatchWriteImageBackingOfTileAttachmentViaNonTileAttachmentApi` | 10675 |
| `DispatchReadImageBackingOfTileAttachmentViaNonTileAttachmentApiAfterDrawWrite` | 10676 |
| `DispatchReadImageBackingOfTileAttachmentViaNonTileAttachmentApiAfterDispatchWrite` | 10676 |
| `DispatchReadFirstElementImageBackingOfTileAttachmentArrayAfterDispatchWrite` | 10676, `ImageArray`, `DynamicRendering` |
| `DispatchReadImageBackingOfTileAttachmentAfterDrawWrite` | 10676, `Positive` |
| `DispatchTileAttachmentArrayWriteFirstOnlyAndReadSecondViaNonTileAttachmentApi` | 10676, `Positive`, `ImageArray`, `DynamicRendering` |
| `DispatchReadFirstElementOfSamplerArrayWhileSecondElementBacksWrittenTileAttachment` | 10676, `Positive`, `ImageArray` |
| `DispatchWriteFirstElementOfStorageImageArrayWhileSecondElementBacksTileAttachment` | 10675, `Positive`, `ImageArray` |
| `DispatchReadInSubpass1ImageWrittenAsTileAttachmentInSubpass0` | 10676, `Positive`, `MultipleSupasses` |

### CODE REVIEW

File: layers/state_tracker/cmd_buffer_state.h [hunk 1]
```diff
@@ -496,6 +496,9 @@ class CommandBuffer : public RefcountedStateObject, public SubStateManager<Comma
 
     std::vector<TensorBarrier> tensor_barriers;
 
+    // Record tile attachments written to in a subpass
+    vvl::unordered_set<const vvl::ImageView *> tile_attachments_written_to;
```
Comment (spencer-lunarg) [inline layers/state_tracker/cmd_buffer_state.h]: nit: Move to be above `active_framebuffer`... want to keep all the subpass / attachment stuff somewhat together

File: layers/state_tracker/shader_module.h [hunk 2]
```diff
@@ -373,6 +373,8 @@ struct VariableBase {
 
     bool IsUntyped() const { return data_type_id != 0; }
 
+    bool IsTileAttachmentStorageClass() const { return storage_class == spv::StorageClass::StorageClassTileAttachmentQCOM; }
```
Comment (spencer-lunarg) [inline layers/state_tracker/shader_module.h]: remove this, want to keep the utils for `VariableBase` to a minimum... just inline this

File: layers/state_tracker/state_tracker.cpp [hunk 3]
```diff
@@ -5402,6 +5418,10 @@ void DeviceState::PostCallRecordCmdDrawMultiIndexedEXT(VkCommandBuffer commandBu
                                                        const RecordObject& record_obj) {
     auto cb_state = GetWrite<CommandBuffer>(commandBuffer);
     cb_state->RecordDraw(record_obj.location);
+    if (cb_state->per_tile_execution_model_enabled && cb_state->active_render_pass &&
```
Comment (spencer-lunarg) [inline layers/state_tracker/state_tracker.cpp]: please move all of these `RecordDrawTileAttachmentsWrittenTo` (and dispatch) calls to `CommandBuffer::RecordDraw()` as we don't want to have to copy-and-paste things like this

File: layers/state_tracker/cmd_buffer_state.cpp [hunk 4]
```diff
@@ -1553,6 +1561,87 @@ void CommandBuffer::RecordTraceRay(const Location& loc) {
     }
 }
 
+// Generic function to handle tile attachment written to for all CmdDraw* type functions when per-tile execution model is enabled
+void CommandBuffer::RecordDrawTileAttachmentsWrittenTo() {
```
Comment (spencer-lunarg) [inline layers/state_tracker/cmd_buffer_state.cpp]: pass in the `RecordDrawTileAttachmentsWrittenTo(const LastBound& last_bound)`

File: layers/state_tracker/cmd_buffer_state.cpp [hunk 5]
```diff
@@ -1553,6 +1561,87 @@ void CommandBuffer::RecordTraceRay(const Location& loc) {
...
+    const LastBound& last_bound_graphics = GetLastBoundGraphics();
+
+    if (active_attachments.empty()) {
+        return;
+    }
+
+    bool has_fragment_shader = false;
```
Comment (spencer-lunarg) [inline layers/state_tracker/cmd_buffer_state.cpp]: check if `LastBound::GetFragmentEntryPoint()` is null

File: layers/state_tracker/cmd_buffer_state.cpp [hunk 6]
```diff
@@ -852,6 +855,8 @@ void CommandBuffer::RecordBeginRenderPass(const VkRenderPassBeginInfo& render_pa
     for (auto& item : sub_states_) {
         item.second->RecordBeginRenderPass(render_pass_begin, subpass_begin_info, loc);
     }
+
+    tile_attachments_written_to.clear();
```
Comment (spencer-lunarg) [inline layers/state_tracker/cmd_buffer_state.cpp:859]: does this not work with dynamic rendering? (`vkCmdBeginRendering`)?

If not, leave a comment where it is declared in the header

If it does, please add a dynamic rendering test (1 negative and 1 positive)

File: layers/state_tracker/cmd_buffer_state.cpp [hunk 7]
```diff
@@ -1553,6 +1561,87 @@ void CommandBuffer::RecordTraceRay(const Location& loc) {
...
+
+            const vvl::DescriptorBinding* binding = descriptor_set->GetBinding(binding_index);
+            if (!binding || binding->descriptor_class != DescriptorClass::Image) {
+                continue;
+            }
+
+            uint32_t count = binding->count;
+            const ImageBinding* image_binding = static_cast<const ImageBinding *>(binding);
+            if (resource_variable.IsArray() && !binding->IsVariableCount()) {
+                count = resource_variable.array_length;
```
Comment (spencer-lunarg) [inline layers/state_tracker/cmd_buffer_state.cpp]: if someone goes

```
// all my samplers in a single array, as apps do
layout(set = 0, binding = 0) sampler s[256];

void main() {
    texture(tex, s[4]);
}
```

it will no mark ALL the samplers as written, but only 1 was actually written to

that is a flaw of `IsWrittenTo()`

```
    // Read/Write here refer to the variable itself. For a buffer this means the memory has been accessed. For an Image, this only
    // means the descriptor itself has been accessed
```

so try writing at test that goes

```
layout(set = 0, binding = 0) sampler s[2];

// First subpass
void main() {
    texture(tex, s[0]);
}

// Second subpass
void main() {
    texture(tex, s[1]);
}
```

and you are going to get a false positive

**Review by spencer-lunarg — CHANGES_REQUESTED** [review]

File: layers/state_tracker/cmd_buffer_state.cpp [hunk 8]
```diff
@@ -1553,6 +1561,87 @@ void CommandBuffer::RecordTraceRay(const Location& loc) {
...
+        }
+    }
+}
+
+// Generic function to handle tile attachment written to for all CmdDispatch* type functions when per-tile execution model is enabled
+void CommandBuffer::RecordDispatchTileAttachmentsWrittenTo() {
+    const LastBound& last_bound_compute = GetLastBoundCompute();
+    const vvl::Pipeline* compute_pipeline = last_bound_compute.pipeline_state;
+    ASSERT_AND_RETURN(compute_pipeline);
+
```
Comment (ZZKQcom) [inline layers/state_tracker/cmd_buffer_state.cpp:1607]: Need add shader object of compute.

File: layers/state_tracker/state_tracker.cpp (hunk 3 above)
Comment (spencer-lunarg) [reply layers/state_tracker/state_tracker.cpp]: (note, I noticed `CmdDrawIndirectByteCount` is wrong/missing... I will fix that, please ignore)

File: layers/state_tracker/cmd_buffer_state.cpp (hunk 6 above)
Comment (ZZKQcom) [reply layers/state_tracker/cmd_buffer_state.cpp:859]: Yes, should work with dynamic rendering.

Fix the issue.

Add negative test `DispatchReadFirstElementImageBackingOfTileAttachmentArrayAfterDispatchWrite` for `VUID-vkCmdDispatch-None-10676`, and positive test `DispatchTileAttachmentArrayWriteFirstOnlyAndReadSecondViaNonTileAttachmentApi`.

File: layers/state_tracker/cmd_buffer_state.cpp (hunk 7 above)
Comment (ZZKQcom) [reply layers/state_tracker/cmd_buffer_state.cpp]: Great suggestion.

`IsWrittenTo()` is per-variable and would mark all elements as written even when only one index is accessed. Fixed this by introducing `image_read_indices` and `image_written_indices` in `ResourceInterfaceVariable`, populated during shader module analysis in `shader_module.cpp`. 
  
All VUID checks now use `image_written_indices.contains(index)` or `image_read_indices.contains(index)`.

Added positive unit tests `DispatchTileAttachmentArrayWriteFirstOnlyAndReadSecondViaNonTileAttachmentApi` (`VUID-10676`) and `DispatchWriteFirstElementOfStorageImageArrayWhileSecondElementBacksTileAttachment` (`VUID-10675`).

Added negative unit tests `DispatchWriteImageBackingOfTileAttachmentViaNonTileAttachmentApi` (`VUID-10675`) and `DispatchReadImageBackingOfTileAttachmentViaNonTileAttachmentApiAfterDrawWrite` (`VUID-10676`) and `DispatchReadImageBackingOfTileAttachmentViaNonTileAttachmentApiAfterDispatchWrite` (`VUID-10676`) and `DispatchReadFirstElementImageBackingOfTileAttachmentArrayAfterDispatchWrite` (`VUID-10676`).

**Comment by ZZKQcom** [comment]
Fixed potential issues according to the suggestions, added more unit tests, such as dynamic rendering, image array and multiple subpasses.

File: layers/state_tracker/cmd_buffer_state.h (hunk 1 above)
Comment (ZZKQcom) [reply layers/state_tracker/cmd_buffer_state.h]: Move the declaration.

File: layers/state_tracker/shader_module.h (hunk 2 above)
Comment (ZZKQcom) [reply layers/state_tracker/shader_module.h]: Remove the helper function.

File: layers/state_tracker/state_tracker.cpp (hunk 3 above)
Comment (ZZKQcom) [reply layers/state_tracker/state_tracker.cpp]: Got it. Moved to `RecordDraw` and `RecordDispatch` functions.

File: layers/state_tracker/cmd_buffer_state.cpp (hunk 4 above)
Comment (ZZKQcom) [reply layers/state_tracker/cmd_buffer_state.cpp]: Fix.

File: layers/state_tracker/cmd_buffer_state.cpp (hunk 5 above)
Comment (ZZKQcom) [reply layers/state_tracker/cmd_buffer_state.cpp]: Refactor the implementation.

File: layers/state_tracker/cmd_buffer_state.cpp (hunk 8 above)
Comment (ZZKQcom) [reply layers/state_tracker/cmd_buffer_state.cpp:1607]: Support compute shader object.

File: layers/state_tracker/shader_module.cpp [hunk 9]
```diff
@@ -2560,6 +2560,13 @@ ResourceInterfaceVariable::ResourceInterfaceVariable(const Module& module_state,
                             write_without_formats_component_count_list.push_back(image_access.texel_component_count);
                         }
                     }
+                    // Track statically known array indices being written
+                    if (!IsArray()) {
+                        image_written_indices.emplace(0);
+                    } else if (image_access.image_access_chain_index != kInvalidValue &&
+                               image_access.image_access_chain_index != kSpecConstant) {
+                        image_written_indices.emplace(image_access.image_access_chain_index);
```
Comment (spencer-lunarg) [inline layers/state_tracker/shader_module.cpp:2568]: how would this work if the use goes

```
layout(set = 0, binding = 0) sampler s[256];

layout(set = 0, binding = 0) uniform UBO {
    uint index;
};

void main() {
    texture(tex, s[index]);
}
```

it won't be able to detect the value here

... I'll just let you know, this is not possible to fully detect on the CPU and **will** need GPU-AV to solve.
This is why we have `DescriptorValidator::ValidateDescriptor` that we use for cases like this when we can't detect the exact index a descriptor is written to. That function will be called only when we know we can validate that variable

... the issue you are running into is that you need to track "previous" read/writes, which is something we plan to add (for other things) but don't have a system currently in place for.

You can't just "store the value" per `resource_variable` either because you could have code that looks like

```c++
if (some_condition) {
    texture(tex, s[index]);
} else {
    texture(tex, s[index + 1]);
}
```

and have multiple draws share the same pipeline (or multiple pipelines share the same SPIR-V), or even multiple command buffers submitted at once... basically you can't have a `OpVariable` (resource_variable) hold state, and we need to somehow build a system where each command buffer has as temporary "view" of what has just got written or not in a shader

... another example is if you go

```
vkCmdDraw();
vkCmdDraw(); // right away, will 100% be invalid as it now read/writes again
``` 

none of this code you have will detect that, you need to state track on the GPU what has been seen

........ SO.... I am not against you trying to add this, but it 100% going to need to be done in GPU-AV and if you add it, since the majority of people using GPU-AV don't plan to use this vendor extension, it needs to be something that performance wise is not noticeable when the extension is not enabled

Comment (spencer-lunarg) [reply layers/state_tracker/shader_module.cpp:2568]: personal opinion, the return on investment to add these 3 VUs is going to be LOW, so please only embark on this if you really are sure you want to support these (there are many core/EXT/KHR VUs like this that are also not supported)

Comment (ZZKQcom) [reply layers/state_tracker/shader_module.cpp:2568]: I understand your point. I will discuss this with Matt, and for the time being, I will mark these VUIDs as TODOs.

**Comment by ZZKQcom** [comment]
Mark these 3 VUIDs as TODOs:
`VUID-vkCmdDispatch-aspectMask-10673`
`VUID-vkCmdDispatch-None-10675`
`VUID-vkCmdDispatch-None-10676`

**Comment by spencer-lunarg** [comment]
@ZZKQcom I marked them in https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/11339

thanks again, I feel there is better validation to spend your energy on than these 3 dispatch VUs, one day hopefully it will be easier to just add them

**Comment by spencer-lunarg** [comment]
@mnetsch @ZZKQcom I am sorting the GPU-AV stuff right now

` VUID-vkCmdDraw-None-09003` is likely what is the most similar to this, will note when we get this added, to try and just grab in these 3 VUs as well
